File: Data Structures & Algorithms/max-area-of-island/submission-1.py
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        rows,cols=len(grid),len(grid[0])
        visited=set()
        max_size=float('-inf')
        def bfs(r,c):
            q=collections.deque()
            q.append((r,c))
            visited.add((r,c))
            count=1

            while q:
                rw,cl=q.pop()
                directions=[[1,0],[-1,0],[0,1],[0,-1]]
                for dr,dc in directions:
                    if 0 <= rw + dr < rows and 0 <= cl + dc < cols and grid[dr+rw][dc+cl]==1 and (dr+rw,dc+cl) not in visited:
                        print("Count:",count)
                        count+=1
                        visited.add((dr+rw,dc+cl))
                        q.append((dr+rw,dc+cl))
                        
                        
            return count


        for r in range(rows):
            for c in range(cols):
                if grid[r][c]==1 and (r,c) not in visited:
                    m=bfs(r,c)
                    if max_size<m:
                        max_size=m
            
        if(max_size==float('-inf')):
            return 0
        else: 
            return max_size
                
                
        

File: Data Structures & Algorithms/count-number-of-islands/submission-1.cpp
#include<iostream>
#include<set>
#include<vector>
#include<deque>
using namespace std;
class Solution {
public:
    set <pair<int,int>> visited;
    void bfs(int r,int c,int rows,int cols,vector<vector<char>>& grid)
        {
            deque<pair<int,int>> q;
            visited.insert({r,c});
            q.push_back({r,c});
            vector <pair<int,int>> directions={
                {1,0},
                {-1,0},
                {0,1},
                {0,-1}
            };
            while (!q.empty())
            {
                auto current=q.front();
                q.pop_front();
                int i=current.first;
                int j=current.second;
                for (auto dir: directions)
                {
                    int dr=dir.first;
                    int dc=dir.second; 
                    int nr=i+dr;
                    int nc=j+dc;
                    if (nr>=0 && nr<rows && nc>=0 && nc<cols && grid[nr][nc]=='1' && !visited.count({nr,nc}))
                    {
                        q.push_back({nr,nc});
                        visited.insert({nr,nc});
                    }
                }
                    

            }
        }
   
    int numIslands(vector<vector<char>>& grid) {
        if (grid.empty() || grid[0].empty()){
            return 0;
        }
        int row=grid.size();
        int col=grid[0].size();
        
        int islands=0;
        
        for (int r=0;r<row;r++) {
            for (int c=0;c<col;c++) {
                if(grid[r][c]=='1' && !visited.count({r,c}))
                {
                    bfs(r,c,row,col,grid);
                    islands++;
                }
            }
        }
        return islands;
    }
};
